pkg/chunk/cached_store.go

```diff
@@ -155,11 +155,16 @@ func (s *rSlice) ReadAt(ctx context.Context, page *Page, off int) (n int, err er
 		}
 		// partial read
 		st := time.Now()
-		in, err := s.store.storage.Get(key, int64(boff), int64(len(p)))
-		if err == nil {
-			n, err = io.ReadFull(in, p)
-			_ = in.Close()
-		}
+		page.Acquire()
+		err := utils.WithTimeout(func() error {
+			defer page.Release()
+			in, err := s.store.storage.Get(key, int64(boff), int64(len(p)))
```

Comment:
goroutines that started by utils.WithTimeout will not be interrupted and continue to execute or block (which is not the desired outcome). and may cause the number of goroutines to keep increasing? maybe we need pass context to object storage interface and cancel it, like:
```golang
type ObjectStorage interface {
	Get(ctx context, key string, off, limit int64, getters ...AttrGetter) (io.ReadCloser, error)
}

// i.e. pass the context instead of context.Background()
func (s *s3client) Get(ctx context, key string, off, limit int64, getters ...AttrGetter) (io.ReadCloser, error) {
     ....
     resp, err := s.s3.GetObjectWithContext(ctx, params, request.WithGetResponseHeader(s3RequestIDKey, &reqID))
}
```

Reply:
we can do this later

Reply:
Agreed, we can do our best effort to support object clients that accept a ctx.

Reply:
issue #4642

Reply:
When ctx parameter is supported, I think we can pass `getters ...AttrGetter` via `context.WithValue`